### DISCUSSION TITLE
Proposal: new index on positions

### DISCUSSION THREAD

**Discussion by morten-holm** [teslamate-org/teslamate#2009] [Ideas]
Okay, so I've noticed that after running Teslamate on my Raspberry Pi 4 for over a year especially the 'Km driven' on the 'Drive Stats' has become sluggish.

I propose the following index:

create index 'name that you pick' on positions ((date_trunc('day', date)::date));

I have done some timings on my Pi for 7 days and 365 days with/without the index, here are the results:

7 days | Before | After |
------| ------| -----|
1st run | 0m52.566s |  0m11.859s |
2nd run | 0m41.985s |  0m11.745s |
3rd  | 0m32.045s | 0m11.647s |

365 days | before | After
---|---|---|
1st run | 0m31.581s | 0m12.100s |
2nd run | 0m21.928s | 0m11.793s |
3rd run | 0m27.231s | 0m11.565s | 

It will improve response times, but I can't tell if/how this will impact logging the positions. The rest of the dashboard is pretty instant and the index will only get it some of the way.

**Comment by leewillis77** [comment]
The fact that both the 7 days and 365 days take ~11s seems strange. Surely the 7 day version should be quicker than the 365 day one?

**Comment by dyxyl** [comment]
@leewillis77 I think it is reading the whole positions table and then filtering it by date afterwards.

**Comment by Dulanic** [comment]
Indexing isn't necessary, the report shouldnt be built off positions as that table is just going to get slower and slower over time. I redid this query to use drives instead.

**Comment by Dulanic** [comment]
My previous comment was wrong after speaking to @adriankumpf . I will look at the indexing and come up /w a batter plan.

**Comment by Dulanic** [comment]
@morten-holm So I dont see the index helping too much, the problem is it needs to scan hundreds of thousands of rows. 

1st, can you test this query in grafana? This should be SLIGHTLY faster.

The other option is 100% up to @adriankumpf and that would be to create a new table where the distance travelled could be kept and he would need to create the code in elixir to do so. Honestly, the only reason it is as slow as it is for you is because of the Rasp Pi hardware. SD cards aren't great for databases.  

```

SELECT
    date::date as time,
    convert_km(COALESCE(max(odometer) - min(odometer), 0)::numeric, '$length_unit') as "distance_$length_unit" 
FROM
    positions
WHERE
    car_id = $car_id
    AND $__timeFilter(date)
GROUP BY
    date::date
ORDER BY
    1
```
